File: ncRegras/rCheque.vb
Imports System.Transactions

Imports ncDados.nsCheques
Imports ncPersistencia.nsCheques
Imports ncComum.nsFuncoes
Imports ncComum.nsExcecao

Namespace nsCheques

  Public Class rCheques

    '-- Métodos de controle ( Várias chamadas; Controle de transação )

    Public Function Listar() As ColecaoCheques

      Dim retorno As ColecaoCheques
      Dim persistencia As pCheques
      Dim retornoPersistencia As ColecaoCheques

      Try

        retorno = New ColecaoCheques

        persistencia = New pCheques
        retornoPersistencia = persistencia.Listar()

        If Not retornoPersistencia Is Nothing Then
          If retornoPersistencia.Count > 0 Then
            retorno.AddRange(retornoPersistencia)
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Listar Cheques [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Listar = retorno

    End Function

    Public Function Consultar(ByVal dados As dCheques) As ColecaoCheques

      Dim retorno As ColecaoCheques
      Dim persistencia As pCheques
      Dim retornoPersistencia As ColecaoCheques

      Try

        retorno = New ColecaoCheques

        persistencia = New pCheques
        retornoPersistencia = persistencia.Consultar(dados)

        If Not retornoPersistencia Is Nothing Then
          If retornoPersistencia.Count > 0 Then
            retorno.AddRange(retornoPersistencia)
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Consultar Cheques [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Consultar = retorno

    End Function

    Public Function ConsultarCheques(ByVal dados As dCheques) As ColecaoCheques

      Dim retorno As ColecaoCheques
      Dim persistencia As pCheques
      Dim retornoPersistencia As ColecaoCheques

      Try

        retorno = New ColecaoCheques

        persistencia = New pCheques
        retornoPersistencia = persistencia.Consultar(dados)

        If Not retornoPersistencia Is Nothing Then
          If retornoPersistencia.Count > 0 Then
            retorno.AddRange(retornoPersistencia)
          Else
            retorno = Nothing
          End If
        Else
          retorno = Nothing
        End If

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Consultar Cheques [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Return retorno

    End Function

    Public Function Incluir(ByVal dadosCheques As ColecaoCheques) As Integer

      Dim retorno As Integer
      Dim persistenciaCheques As pCheques

      Try

        Using ts As New TransactionScope

          persistenciaCheques = New pCheques()
          For Each cheques As dCheques In dadosCheques
            persistenciaCheques.Incluir(cheques)
          Next

          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Incluir Cheques [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Incluir = retorno

    End Function

    Public Function Incluir(ByVal dados As dCheques) As Integer

      Dim retorno As Integer
      Dim persistenciaCheques As pCheques

      Try

        Using ts As New TransactionScope

          persistenciaCheques = New pCheques()
          persistenciaCheques.Incluir(dados)
          
          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Incluir Cheques [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Incluir = retorno

    End Function

    Public Function Alterar(ByVal dados As dCheques, ByVal dadosCheques As ColecaoCheques) As Integer

      Dim retorno As Integer
      Dim persistencia As pCheques
      Dim persistenciaCheques As pCheques

      Try

        Using ts As New TransactionScope
          persistencia = New pCheques
          retorno = persistencia.Alterar(dados)

          persistenciaCheques = New pCheques()
          For Each Cheques As dCheques In dadosCheques
            persistenciaCheques.Alterar(Cheques)
          Next


          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Alterar Cheques [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Alterar = retorno

    End Function

    Public Function Baixar() As Integer

      Dim retorno As Integer
      Dim persistencia As pCheques

      Try

        Using ts As New TransactionScope
          persistencia = New pCheques
          retorno = persistencia.Baixar()

          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Baixar Cheques [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Return retorno

    End Function
    Public Function Excluir(ByVal dados As dCheques) As Integer

      Dim retorno As Integer
      Dim persistencia As pCheques

      Try

        Using ts As New TransactionScope
          persistencia = New pCheques
          retorno = persistencia.Excluir(dados)
          ts.Complete()
        End Using

      Catch ex As Exception

        retorno = Nothing
        Throw New ExcecaoNascomercio("Erro em Excluir Cheques [" & Me.ToString() & "] - " & ex.Message)

      End Try

      Excluir = retorno

    End Function

  End Class

End Namespace

File: Servicos/rCheque.cs
using System;
using Comum;
using Repositorios;
using Modelos;

namespace Servicos.nsCheques
{
    public class rCheque : IsCheque
    {
        private readonly IpCheques _repo;
        public rCheque(IpCheques repo) { _repo = repo; }

        public ColecaoCheques Listar()
        {
            try { return _repo.Listar(); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Listar Cheque [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoCheques Consultar(dCheques dados)
        {
            try { return _repo.Consultar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Consultar Cheque [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoCheques ConsultarCheques(dCheques dados)
        {
            try { return _repo.Consultar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarCheques Cheque [" + ToString() + "] - " + ex.Message); }
        }

        public int Incluir(ColecaoCheques dadosCheques)
        {
            try
            {
                foreach (var cheque in dadosCheques)
                    _repo.Incluir(cheque);
                return 1;
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Incluir Cheque [" + ToString() + "] - " + ex.Message); }
        }

        public int Incluir(dCheques dados)
        {
            try { return _repo.Incluir(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Incluir Cheque [" + ToString() + "] - " + ex.Message); }
        }

        public int Alterar(dCheques dados, ColecaoCheques dadosCheques)
        {
            try
            {
                _repo.Alterar(dados);
                foreach (var cheque in dadosCheques)
                    _repo.Alterar(cheque);
                return 1;
            }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Alterar Cheque [" + ToString() + "] - " + ex.Message); }
        }

        public int Baixar()
        {
            try { return _repo.Baixar(); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Baixar Cheque [" + ToString() + "] - " + ex.Message); }
        }

        public int Excluir(dCheques dados)
        {
            try { return _repo.Excluir(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Excluir Cheque [" + ToString() + "] - " + ex.Message); }
        }
    }
}

